torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -591,6 +592,28 @@ ProcessGroupNCCL::ProcessGroupNCCL(
             << "\nUSE_HIGH_PRIORITY_STREAM: "
             << options_->is_high_priority_stream
             << "\nNCCL_DEBUG: " << ncclDebugLevel;
+
+  ucc_lib = loadTorchUCC();
+  if (ucc_lib) {
+    LOG(INFO) << "[Rank " << rank_  << "] torch_ucc.so loaded";
+    typedef void *fn(void *);
+    auto createProcessGroupUCCForNCCL = reinterpret_cast<fn*>(ucc_lib->sym("_Z28createProcessGroupUCCForNCCLPv"));
+    struct args_t {
+      const c10::intrusive_ptr<Store>& store;
+      int rank = -1;
+      int size = -1;
+    };
+    args_t args {
+      .store = store,
+      .rank = rank,
+      .size = size
+    };
+    auto raw_ucc_pg = static_cast<ProcessGroup *>(createProcessGroupUCCForNCCL(&args));
+    ucc_pg = c10::intrusive_ptr<ProcessGroup>::unsafe_steal_from_new(raw_ucc_pg);
```

Comment:
Because PyTorch doesn't know the definition of `class ProcessGroupUCC`, there is no way to create an instance of that object in PyTorch. But it is possible to use `intrusive_ptr::make` to create a `c10::intrusive_ptr<ProcessGroup>` in `torch_ucc.so` and return it to PyTorch. The only problem with this approach is the mangled name will be very long and unreadable.
For example, this function
```C++
c10::intrusive_ptr<c10d::ProcessGroup> createProcessGroupUCC(
  const c10::intrusive_ptr<c10d::Store>& store,
  tagging::world_size_t rank,
  tagging::world_size_t size);
```
has mangled name
```C++
_Z21createProcessGroupUCCRKN3c1013intrusive_ptrIN4c10d5StoreENS_6detail34intrusive_target_default_null_typeIS2_EEEEii
```
I intentionally made the mangled name simple, that's why I define `createProcessGroupUCCForNCCL` as
```C++
void *createProcessGroupUCCForNCCL(void *args)
```